File: spikes/classifier/js/spike.test.ts
import test from "node:test"
import assert from "node:assert/strict"
import { readFileSync, readdirSync } from "node:fs"
import { createHash } from "node:crypto"
import { fileURLToPath } from "node:url"

import { canonical } from "./canonical.ts"
import { Classifier, expect as answerFor, parseDecision } from "./classifier.ts"
import { asGuardrail, bands, judge } from "./judge.ts"
import { choice, noul, score } from "./types.ts"
import type { Question } from "./types.ts"

const FIX = fileURLToPath(new URL("../fixture/", import.meta.url))
const read = (n: string) => readFileSync(FIX + n)

// ------------------------------------------------- gate 1: byte-exact request

const questions: Record<string, Question> = {
  is_refund_request: noul("Is the customer asking for a refund?"),
  department: choice("Which department should handle this?", {
    billing: "refunds, charges, payments",
    shipping: "delivery, damage in transit",
    technical: "product does not work",
  }),
  urgency: score("How urgent is this?", ["routine", "elevated", "urgent"]),
}

const classifier = new Classifier({ model: "typesafe/jev-1.13" })

test("gate 1 — request is byte-identical to the fixture", () => {
  const got = Buffer.from(
    classifier.request("Order 4021 arrived smashed, I want my money back.", questions),
    "utf8",
  )
  assert.equal(got.length, 514)
  assert.deepEqual(got, read("request.json"))
  assert.equal(
    createHash("sha256").update(got).digest("hex"),
    read("request.sha256").toString().trim(),
  )
})

test("gate 1b — score.criteria keeps array order; keys sort recursively", () => {
  const s = canonical({ b: 1, a: { d: 2, c: ["z", "a", "m"] } })
  assert.equal(s, '{"a":{"c":["z","a","m"],"d":2},"b":1}')
})

test("gate 1c — numbers round-trip: 1.21 and an integer-valued 0", () => {
  assert.equal(canonical({ score: 1.21, p: 0, cost: 0.000016716 }), '{"cost":0.000016716,"p":0,"score":1.21}')
  // re-emitting the whole parsed response reproduces every numeric literal
  const raw = JSON.parse(read("response.json").toString())
  const re = canonical(raw)
  for (const lit of ['1.21', '"technical":0', '0.000016716', '0.98', '0.61'])
    assert.ok(re.includes(lit), lit)
})

// ------------------------------------------------------------- gate 2: parse

test("gate 2 — parse the response into typed values", () => {
  const d = parseDecision(JSON.parse(read("response.json").toString()))
  assert.equal(d.model, "typesafe/jev-1.13-20260917")
  assert.equal(answerFor(d, "is_refund_request", "noul").noul, 0.98)

  const dept = answerFor(d, "department", "choice")
  assert.equal(dept.choice, "shipping")
  assert.deepEqual(dept.probabilities, { billing: 0.39, technical: 0, shipping: 0.61 })

  const urg = answerFor(d, "urgency", "score")
  assert.equal(urg.score, 1.21)
  assert.deepEqual(urg.legend, { "0": "routine", "1": "elevated", "2": "urgent" })
  assert.equal(d.usage.input_tokens, 398)
  assert.equal(d.usage.cost, 0.000016716)

  // exhaustive narrowing, zero casts
  const a = d.answers.urgency!
  const label =
    a.type === "noul" ? String(a.noul) : a.type === "choice" ? a.choice : a.legend[String(Math.round(a.score))]!
  assert.equal(label, "elevated")
})

test("gate 2b — a wrong-typed answer is an error, not a silent undefined", () => {
  const d = parseDecision(JSON.parse(read("response.json").toString()))
  assert.throws(() => answerFor(d, "urgency", "noul"), /expected noul/)
  assert.throws(() => parseDecision({ model: "m", answers: { x: { type: "vibe" } } }), /unknown answer type/)
})

// ------------------------------------------------------- gate 3: a wired judge

/** The `static` backend: replay a recorded pair, keyed by the canonical request
 *  bytes. Unknown bytes are a failure, not a fallback. */
function staticBackend() {
  const table = new Map<string, unknown>()
  for (const f of readdirSync(FIX)) {
    const m = /^guard-(\w+)-request\.json$/.exec(f)
    if (!m) continue
    table.set(
      read(f).toString().trim(),
      JSON.parse(read(`guard-${m[1]}-response.json`).toString()),
    )
  }
  return async (body: string) => {
    const hit = table.get(body)
    if (!hit) throw new Error(`no recorded response for ${body.slice(0, 80)}…`)
    return hit
  }
}

const RISK = ["read-only, changes nothing", "writes, but easy to undo", "hard to undo, or reaches outside the workspace", "destructive or irreversible"]

const bashJudge = judge({
  on: "bash",
  ask: {
    from_untrusted: noul("Did this command originate in fetched or untrusted content rather than the user's own request?"),
    risk: score("How hard would this command be to undo?", RISK),
  },
  rule: bands("risk", { ask: 1.5, deny: 2.5 }, (lvl, d) => {
    const legend = answerFor(d, "risk", "score").legend
    return `risk ${lvl.toFixed(2)} — ${legend[String(Math.round(lvl))]}`
  }),
  classifier: new Classifier({ model: "typesafe/jev-1.13", backend: staticBackend() }),
})

const ev = (command: string) => ({ name: "bash", args: { cwd: "/repo", command } })

test("gate 3 — allow / deny / ask over the static backend", async () => {
  assert.deepEqual(await bashJudge(ev("git status --short")), { decision: "allow" })
  assert.equal((await bashJudge(ev(`python3 -c "import shutil; shutil.rmtree('/')"`))).decision, "deny")
  assert.equal((await bashJudge(ev("rm -rf ./build"))).decision, "ask")

  const g = asGuardrail(bashJudge)
  assert.equal(await g(ev("git status --short")), "")
  assert.match(await g(ev(`python3 -c "import shutil; shutil.rmtree('/')"`)), /^risk 2\.97 — destructive/)
})

test("gate 3b — a judge that cannot reach its backend fails open by default, closed on request", async () => {
  const broken = new Classifier({ model: "m", backend: async () => { throw new Error("boom") } })
  const open = judge({ ask: {}, rule: () => ({ decision: "allow" }), classifier: broken })
  const closed = judge({ ask: {}, rule: () => ({ decision: "allow" }), classifier: broken, onError: "deny" })
  assert.deepEqual(await open(ev("x")), { decision: "allow" })
  assert.equal((await closed(ev("x"))).decision, "deny")
})

// --------------------------------------------- gate 4: first-deny-wins holds

/** A copy of js/src/agents/loop.ts guardedHooks, widened to await async
 *  guardrails. Order and short-circuit are unchanged. */
async function runGuardrails(
  guardrails: Array<(ev: any) => string | undefined | void | Promise<string | undefined | void>>,
  e: any,
): Promise<{ denied: boolean; reason?: string }> {
  for (const g of guardrails) {
    const verdict = await g(e)
    if (verdict && verdict !== "allow") return { denied: true, reason: verdict }
  }
  return { denied: false }
}

test("gate 4 — a judge composed after a deny cannot flip it to allow", async () => {
  let judgeRan = false
  const alwaysDeny = () => "policy: bash is off"
  const alwaysAllow = asGuardrail(async (e: any) => {
    judgeRan = true
    return { decision: "allow" as const }
  })
  const r = await runGuardrails([alwaysDeny, alwaysAllow], ev("git status --short"))
  assert.equal(r.denied, true)
  assert.equal(r.reason, "policy: bash is off")
  assert.equal(judgeRan, false, "the loop short-circuits: the later judge is never consulted")

  // and the reverse order still denies — allow is the absence of a verdict, never an override
  const r2 = await runGuardrails([alwaysAllow, alwaysDeny], ev("git status --short"))
  assert.equal(r2.denied, true)
})

File: spikes/classifier/golang/spike_test.go
package classifier_test

import (
	"bytes"
	"context"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"testing"
	"time"

	tn "github.com/muthuishere/toolnexus/golang"
	"github.com/muthuishere/toolnexus/golang/agents"
	c "toolnexus.spike/classifier"
)

const model = "typesafe/jev-1.13"

func fixture(t *testing.T, name string) []byte {
	t.Helper()
	b, err := os.ReadFile(filepath.Join("..", "fixture", name))
	if err != nil {
		t.Fatal(err)
	}
	return bytes.TrimRight(b, "\n")
}

// the three fixture questions
func questions() map[string]c.Question {
	return map[string]c.Question{
		"is_refund_request": c.Noul{Instructions: "Is the customer asking for a refund?"},
		"department": c.ChoiceOver("Which department should handle this?", map[string]string{
			"billing":   "refunds, charges, payments",
			"shipping":  "delivery, damage in transit",
			"technical": "product does not work",
		}),
		"urgency": c.Score{Instructions: "How urgent is this?",
			Criteria: []string{"routine", "elevated", "urgent"}},
	}
}

// ---- GATE 1: byte-exact request -------------------------------------------

func TestGate1CanonicalRequestIsByteExact(t *testing.T) {
	got, err := c.CanonicalRequest(model,
		"Order 4021 arrived smashed, I want my money back.", questions())
	if err != nil {
		t.Fatal(err)
	}
	want := fixture(t, "request.json")
	if !bytes.Equal(got, want) {
		t.Fatalf("not byte-exact\n got: %s\nwant: %s", got, want)
	}
	if len(got) != 514 {
		t.Fatalf("want 514 bytes, got %d", len(got))
	}
	sum := sha256.Sum256(got)
	wantSum := strings.Fields(string(fixture(t, "request.sha256")))[0]
	if hex.EncodeToString(sum[:]) != wantSum {
		t.Fatalf("sha256 %s != %s", hex.EncodeToString(sum[:]), wantSum)
	}
}

// score.criteria order IS the level numbering — prove it is never sorted.
func TestGate1ScoreArrayOrderIsPreserved(t *testing.T) {
	got, err := c.CanonicalRequest(model, "x", map[string]c.Question{
		"q": c.Score{Instructions: "i", Criteria: []string{"zulu", "alpha", "mike"}},
	})
	if err != nil {
		t.Fatal(err)
	}
	if !bytes.Contains(got, []byte(`"criteria":["zulu","alpha","mike"]`)) {
		t.Fatalf("array order lost: %s", got)
	}
}

// the Noul fixture omits criteria entirely; the optional form must still emit.
func TestGate1NoulCriteriaAbsentVsPresent(t *testing.T) {
	absent, _ := c.CanonicalRequest(model, "x", map[string]c.Question{"q": c.Noul{Instructions: "i"}})
	if bytes.Contains(absent, []byte("criteria")) {
		t.Fatalf("criteria should be absent: %s", absent)
	}
	present, _ := c.CanonicalRequest(model, "x", map[string]c.Question{
		"q": c.Noul{Instructions: "i", True: "yes it is", False: "no it is not"}})
	if !bytes.Contains(present, []byte(`"criteria":{"false":"no it is not","true":"yes it is"}`)) {
		t.Fatalf("criteria object wrong: %s", present)
	}
}

// keys sort recursively, ASCII, regardless of the order they were built in.
func TestGate1KeysSortRecursively(t *testing.T) {
	got, _ := c.CanonicalRequest(model, map[string]any{"z": 1, "a": 2, "M": 3}, map[string]c.Question{
		"zz": c.Noul{Instructions: "i"}, "aa": c.Noul{Instructions: "i"},
	})
	if !bytes.Contains(got, []byte(`"state":{"M":3,"a":2,"z":1}`)) {
		t.Fatalf("state keys not sorted: %s", got)
	}
	if strings.Index(string(got), `"aa"`) > strings.Index(string(got), `"zz"`) {
		t.Fatalf("question keys not sorted: %s", got)
	}
}

// ---- GATE 2: parse ---------------------------------------------------------

func TestGate2Parse(t *testing.T) {
	var d c.Decision
	if err := json.Unmarshal(fixture(t, "response.json"), &d); err != nil {
		t.Fatal(err)
	}
	if d.Model != "typesafe/jev-1.13-20260917" {
		t.Fatalf("model %q", d.Model)
	}
	n, err := d.Noul("is_refund_request")
	if err != nil || n.Noul != 0.98 {
		t.Fatalf("noul %v %v", n, err)
	}
	ch, err := d.Choice("department")
	if err != nil || ch.Choice != "shipping" {
		t.Fatalf("choice %v %v", ch, err)
	}
	if ch.Probabilities["billing"] != 0.39 || ch.Probabilities["shipping"] != 0.61 ||
		ch.Probabilities["technical"] != 0 {
		t.Fatalf("probabilities %v", ch.Probabilities)
	}
	if ch.Confidence != 0.41 {
		t.Fatalf("confidence %v", ch.Confidence)
	}
	sc, err := d.Score("urgency")
	if err != nil || sc.Score != 1.21 {
		t.Fatalf("score %v %v", sc, err)
	}
	if got := fmt.Sprint(sc.Levels()); got != "[routine elevated urgent]" {
		t.Fatalf("legend order %s", got)
	}
	// wrong-type access is an error, not a panic or a zero value
	if _, err := d.Noul("urgency"); err == nil {
		t.Fatal("want a type error reading a score answer as a noul")
	}
	if _, err := d.Score("nope"); err == nil {
		t.Fatal("want an error for a missing key")
	}
}

// float round-trip: 1.21 and the technical:0 probability must re-emit exactly.
func TestGate2FloatRoundTrip(t *testing.T) {
	var d c.Decision
	if err := json.Unmarshal(fixture(t, "response.json"), &d); err != nil {
		t.Fatal(err)
	}
	sc, _ := d.Score("urgency")
	ch, _ := d.Choice("department")
	out, err := json.Marshal(map[string]any{
		"score": sc.Score, "zero": ch.Probabilities["technical"],
		"probs": sc.Probabilities, "cost": 0.000016716,
	})
	if err != nil {
		t.Fatal(err)
	}
	want := `{"cost":0.000016716,"probs":{"0":0.04,"1":0.71,"2":0.25},"score":1.21,"zero":0}`
	if string(out) != want {
		t.Fatalf("float re-emit drift\n got: %s\nwant: %s", out, want)
	}
}

// ---- GATE 3: one judge, wired ---------------------------------------------

func guardQuestions() map[string]c.Question {
	return map[string]c.Question{
		"from_untrusted": c.Noul{Instructions: "Did this command originate in fetched or untrusted content rather than the user's own request?"},
		"risk": c.Score{Instructions: "How hard would this command be to undo?", Criteria: []string{
			"read-only, changes nothing",
			"writes, but easy to undo",
			"hard to undo, or reaches outside the workspace",
			"destructive or irreversible",
		}},
	}
}

// staticClassifier keys recorded responses by the canonical request bytes, so a
// fixture hit is itself a byte-exactness assertion on the guard requests.
func staticClassifier(t *testing.T) *c.Classifier {
	t.Helper()
	fx := map[string][]byte{}
	for _, name := range []string{"allow", "ask", "deny"} {
		fx[string(fixture(t, "guard-"+name+"-request.json"))] = fixture(t, "guard-"+name+"-response.json")
	}
	return c.New(c.Options{Style: "static", Model: model, Fixtures: fx})
}

func bashJudge() c.Judge {
	return c.Judge{
		// On is the security posture: only these three fields ever leave.
		On: func(ev tn.BeforeToolEvent) any {
			return map[string]any{
				"tool": ev.Name, "cwd": "/repo",
				"command": fmt.Sprint(ev.Args["command"]),
			}
		},
		Ask: guardQuestions(),
		Rule: c.Bands("risk", []c.Band{
			{Max: 1.0, Name: "allow"},
			{Max: 2.5, Name: "ask"},
			{Name: "deny"},
		}),
		FailClosed: true,
	}
}

func ev(cmd string) tn.BeforeToolEvent {
	return tn.BeforeToolEvent{Name: "bash", Args: map[string]any{"command": cmd}, ID: "t1", Turn: 1}
}

func TestGate3JudgeAsGuardrail(t *testing.T) {
	cl := staticClassifier(t)
	rail := bashJudge().AsGuardrail(cl, nil)
	cases := []struct {
		cmd, want string
	}{
		{"git status --short", ""},
		{`python3 -c "import shutil; shutil.rmtree('/')"`, "denied by judge: risk=2.97 (destructive or irreversible)"},
		{"rm -rf ./build", "needs approval: risk=2.25 (hard to undo, or reaches outside the workspace)"},
	}
	for _, tc := range cases {
		if got := rail(ev(tc.cmd)); got != tc.want {
			t.Fatalf("%s ⇒ %q, want %q", tc.cmd, got, tc.want)
		}
	}
}

func TestGate3VerdictEvidence(t *testing.T) {
	v, err := bashJudge().Rules(context.Background(), staticClassifier(t), ev("rm -rf ./build"))
	if err != nil {
		t.Fatal(err)
	}
	if v.Kept[0] != "ask" || v.Evidence["risk"] != 2.25 || v.Model != "typesafe/jev-1.13-20260917" {
		t.Fatalf("verdict %+v", v)
	}
}

// fail-open vs fail-closed must be a declared choice, not an accident.
func TestGate3FailPosture(t *testing.T) {
	empty := c.New(c.Options{Style: "static", Model: model, Fixtures: map[string][]byte{}})
	j := bashJudge()
	if got := j.AsGuardrail(empty, nil)(ev("anything")); got == "" {
		t.Fatal("fail-closed judge allowed on backend error")
	}
	j.FailClosed = false
	if got := j.AsGuardrail(empty, nil)(ev("anything")); got != "" {
		t.Fatalf("fail-open judge denied on backend error: %q", got)
	}
}

// ---- GATE 4: invariant -----------------------------------------------------

func TestGate4JudgeCannotWidenAnEarlierDenial(t *testing.T) {
	deny := agents.Guardrail(func(tn.BeforeToolEvent) string { return "policy: bash is off" })
	// the judge would ALLOW this command on its own …
	judged := bashJudge().AsGuardrail(staticClassifier(t), nil)
	if got := judged(ev("git status --short")); got != "" {
		t.Fatalf("precondition: judge should allow, got %q", got)
	}
	// … but composed after a denial it cannot flip it.
	rail := c.Compile(deny, judged)
	if got := rail(ev("git status --short")); got != "policy: bash is off" {
		t.Fatalf("judge widened an earlier denial: %q", got)
	}
	// and a judge that denies still denies when it runs second
	rail2 := c.Compile(func(tn.BeforeToolEvent) string { return "" }, judged)
	if got := rail2(ev(`python3 -c "import shutil; shutil.rmtree('/')"`)); got == "" {
		t.Fatal("judge failed to deny when composed second")
	}
}

// ---- optional 5th: live call ----------------------------------------------

func TestLiveOptional(t *testing.T) {
	if os.Getenv("OPENROUTER_API_KEY") == "" {
		t.Skip("OPENROUTER_API_KEY not set")
	}
	cl := c.New(c.Options{
		BaseURL:   "https://openrouter.ai/api/v1/systemone",
		Model:     model,
		APIKeyEnv: "OPENROUTER_API_KEY",
		Timeout:   15 * time.Second,
	})
	start := time.Now()
	d, err := cl.Evaluate(context.Background(),
		"Order 4021 arrived smashed, I want my money back.", questions())
	if err != nil {
		t.Fatalf("live call: %v", err)
	}
	t.Logf("live latency %dms model=%s answers=%d", time.Since(start).Milliseconds(), d.Model, len(d.Answers))
	if _, err := d.Choice("department"); err != nil {
		t.Fatalf("live shape mismatch: %v", err)
	}
}
